onnxruntime/test/testdata/custom_op_local_function/custom_gemm.cc

```diff
@@ -0,0 +1,792 @@
+  for (int64_t i = 0; i < n; ++i) {
+    dst[i] = e4m3fn_to_float(src[i]);
+  }
+}
+
+void e4m3fn_to_float(int64_t n, const uint8_t* src, float* dst, float scale) {
+  for (int64_t i = 0; i < n; ++i) {
+    dst[i] = e4m3fn_to_float(src[i]) * scale;
+  }
+}
+
+//////////////////
+// CustomGemmOp...
+//////////////////
+
+void* CustomGemmOp::CreateKernel(const OrtApi& api,
+                                 const OrtKernelInfo* info) const {
+  return std::make_unique<CustomGemmKernel>(api, info).release();
+}
+
+const char* CustomGemmOp::GetName() const { return op_name_; }
+
+const char* CustomGemmOp::GetExecutionProviderType() const {
+  return "CPUExecutionProvider";
+}
+
+size_t CustomGemmOp::GetInputTypeCount() const { return 6; }
+
+ONNXTensorElementDataType CustomGemmOp::GetInputType(size_t index) const {
+  switch (index) {
+    case 0:  // A
+    case 1:  // B
+      return ab_type_;
+    case 2:  // C
+      return c_type_;
+    case 3:  // scale A
+    case 4:  // scale B
+    case 5:  // scale Y
+      return ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT;
+    default:
+      ORT_CXX_API_THROW("Input index is out of boundary.", OrtErrorCode::ORT_RUNTIME_EXCEPTION);
+  }
+}
+
+OrtCustomOpInputOutputCharacteristic
+CustomGemmOp::GetInputCharacteristic(size_t index) const {
+  switch (index) {
+    case 0:
+    case 1:
+      return OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_REQUIRED;
+    case 2:
+      return OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_OPTIONAL;
+    case 3:
+    case 4:
+    case 5:
+      return OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_OPTIONAL;
+    default:
+      ORT_CXX_API_THROW("Input index is out of boundary.", OrtErrorCode::ORT_RUNTIME_EXCEPTION);
+  }
+}
+
+size_t CustomGemmOp::GetOutputTypeCount() const { return 1; }
+
+ONNXTensorElementDataType CustomGemmOp::GetOutputType(size_t index) const {
+  // D, scale D
+  switch (index) {
+    case 0:
+      return d_type_;
+    default:
+      ORT_CXX_API_THROW("Output index is out of boundary.", OrtErrorCode::ORT_RUNTIME_EXCEPTION);
+  }
+}
+
+OrtCustomOpInputOutputCharacteristic
+CustomGemmOp::GetOutputCharacteristic(size_t index) const {
+  switch (index) {
+    case 0:
+      return OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_REQUIRED;
+    default:
+      ORT_CXX_API_THROW("Output index is out of boundary.", OrtErrorCode::ORT_RUNTIME_EXCEPTION);
+  }
+}
+
+///////////////////
+// CustomGemmKernel
+///////////////////
+
+CustomGemmKernel::CustomGemmKernel(const OrtApi& api,
+                                   const OrtKernelInfo* info) {
+  rowMajor_ = KernelInfoGetOptionalAttribute<int64_t>(api, info, "rowMajor", 1);
+  transA_ =
+      KernelInfoGetOptionalAttributeInt64AsBool(api, info, "transA", false);
+  transB_ =
+      KernelInfoGetOptionalAttributeInt64AsBool(api, info, "transB", false);
+  fastAccumulationMode_ = KernelInfoGetOptionalAttributeInt64AsBool(
+      api, info, "fastAccumulationMode", true);
```

Comment:
nit: readability will be improved by using the 120 chars allowed

Reply:
removed